libzmix/src/verifiable_encryption/cs_verifiable_encryption.rs

```diff
@@ -0,0 +1,849 @@
+// Copyright contributors to Hyperledger Ursa.
+// SPDX-License-Identifier: Apache-2.0
+
+/// Camenisch-Shoup verifiable encryption.
+/// Based on the paper Practical Verifiable Encryption and Decryption of Discrete Logarithms. https://www.shoup.net/papers/verenc.pdf
+/// Need to be used with Anonymous credentials as described in Specification of the Identity
+/// Mixer Cryptographic Library, https://domino.research.ibm.com/library/cyberdig.nsf/papers/EEB54FF3B91C1D648525759B004FBBB1/$File/rz3730_revised.pdf
+/// sections 5.3, 6.2.10 and 6.2.19
+use super::bn::{BigNumber, BigNumberContext, BIGNUMBER_1, BIGNUMBER_2};
+use super::errors::prelude::*;
+
+use super::cl::constants::*;
+use super::cl::hash::get_hash_as_int;
+use super::cl::helpers::*;
+
+#[derive(Serialize, Deserialize)]
+pub struct PaillierGroup {
+    pub g: BigNumber,
+    pub h: BigNumber,
+    pub n_by_4: BigNumber, // n/4, precomputation
+    pub modulus: BigNumber,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct CSEncPrikey {
+    pub x1: Vec<BigNumber>,
+    pub x2: BigNumber,
+    pub x3: BigNumber,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct CSEncPubkey {
+    pub n: BigNumber,
+    pub two_inv_times_2: BigNumber, // (2^-1 % n) * 2, precomputation
+    pub paillier_group: PaillierGroup,
+    pub y1: Vec<BigNumber>,
+    pub y2: BigNumber,
+    pub y3: BigNumber,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct CSCiphertext {
+    pub u: BigNumber,
+    pub e: Vec<BigNumber>,
+    pub v: BigNumber,
+}
+
+pub struct CSKeypair {
+    pub pri_key: CSEncPrikey,
+    pub pub_key: CSEncPubkey,
+}
+
+impl PaillierGroup {
+    /// Order (modulus) is n^2
+    pub fn new(n: &BigNumber, ctx: &mut BigNumberContext) -> UrsaCryptoResult<Self> {
```

Comment:
This isn't "rusty". `new` is templated to take no arguments. A better practice is to use a name like `new_with_modulus`

Reply:
This does seem reasonable. Can you point me to a guideline that says `new` should not take any arguments? Secondly, lot of places in Ursa have `new` with arguments.

Reply:
run `cargo clippy` it will generate a warning for this.

Reply:
What flags are you running with? Which version of clippy? I don't see that warning.

Reply:
`cargo clippy --all -- -W clippy::not_unsafe_ptr_arg_deref -A clippy::many_single_char_names`

Reply:
Don't see the warning? Can you paste the warning that you get here?